{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d012a9cc-3472-4493-b17a-2f6c131158a6",
   "metadata": {},
   "source": [
    "## Strain R21.0464"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac1a991e-b610-4d1d-abbd-7cee900288d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "LR = '/media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/barcode03.fastq.gz'\n",
    "SR1 = '/media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/R20.0464_R1.fastq.gz'\n",
    "SR2 = '/media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/R20.0464_R2.fastq.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a32a98b-395f-4386-b398-44af81adae3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "403d6934-a871-4a59-bcb4-143714908716",
   "metadata": {},
   "outputs": [],
   "source": [
    "outpath = Path('/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8f472914-6da5-45ce-a9c7-55d6e6878243",
   "metadata": {},
   "outputs": [],
   "source": [
    "subsample_dirname = outpath/'subsample'\n",
    "subsample_dirname.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b802f51b-22df-4964-a9d2-3ab0eaa86380",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler read subsampling\u001b[0m \u001b[2m(2022-01-27 13:56:02)\u001b[0m\n",
      "\u001b[2m    Trycycler subsample is a tool for subsampling a long-read set into subsets\u001b[0m\n",
      "\u001b[2mthat are maximally independent from each other.\u001b[0m\n",
      "\n",
      "Output directory already exists: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample\n",
      "\n",
      "Input reads: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/barcode03.fastq.gz\n",
      "  50,065 reads (586,284,896 bp)\n",
      "  N50 = 20,749 bp\n",
      "\n",
      "Checking required software:\n",
      "  minimap2: v2.22-r1101\n",
      "  miniasm: v0.3-r179\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCalculating subset size\u001b[0m \u001b[2m(2022-01-27 13:56:15)\u001b[0m\n",
      "\u001b[2m    Trycycler will now calculate the number of reads to put in each subset.\u001b[0m\n",
      "\n",
      "Total read depth: 122.1x\n",
      "Mean read length: 11,710 bp\n",
      "\n",
      "Calculating subset sizes:\n",
      "  subset_depth = 25.0 * log_2(4 * total_depth / 25.0) / 2\n",
      "               = 53.6x\n",
      "  reads per subset: 21,973\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mSubsetting reads\u001b[0m \u001b[2m(2022-01-27 13:56:15)\u001b[0m\n",
      "\u001b[2m    This step shuffles the reads and saves them into subset files.\u001b[0m\n",
      "\n",
      "Shuffling reads... done\n",
      "\n",
      "subset 0:\n",
      "  reads 1-21973\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_01.fastq\n",
      "  257,460,412 bp\n",
      "\n",
      "subset 1:\n",
      "  reads 4173-26145\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_02.fastq\n",
      "  257,807,206 bp\n",
      "\n",
      "subset 2:\n",
      "  reads 8345-30317\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_03.fastq\n",
      "  256,456,326 bp\n",
      "\n",
      "subset 3:\n",
      "  reads 12517-34489\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_04.fastq\n",
      "  257,030,696 bp\n",
      "\n",
      "subset 4:\n",
      "  reads 16689-38661\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_05.fastq\n",
      "  255,873,668 bp\n",
      "\n",
      "subset 5:\n",
      "  reads 20861-42833\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_06.fastq\n",
      "  257,168,564 bp\n",
      "\n",
      "subset 6:\n",
      "  reads 25033-47005\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_07.fastq\n",
      "  257,000,580 bp\n",
      "\n",
      "subset 7:\n",
      "  reads 29206-50065 and 1-1113\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_08.fastq\n",
      "  256,341,469 bp\n",
      "\n",
      "subset 8:\n",
      "  reads 33378-50065 and 1-5285\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_09.fastq\n",
      "  257,011,533 bp\n",
      "\n",
      "subset 9:\n",
      "  reads 37550-50065 and 1-9457\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_10.fastq\n",
      "  258,189,671 bp\n",
      "\n",
      "subset 10:\n",
      "  reads 41722-50065 and 1-13629\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_11.fastq\n",
      "  259,033,851 bp\n",
      "\n",
      "subset 11:\n",
      "  reads 45894-50065 and 1-17801\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/subsample/sample_12.fastq\n",
      "  258,059,057 bp\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinished!\u001b[0m \u001b[2m(2022-01-27 13:59:30)\u001b[0m\n",
      "\u001b[2m    Now you must now assemble each of the subsampled read sets to produce a set\u001b[0m\n",
      "\u001b[2mof assemblies you can input into Trycycler cluster.\u001b[0m\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler subsample -r {LR} -o {subsample_dirname} --genome_size 4.8M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2cdc5998-95a3-431d-9416-88946bf6f118",
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblies_dirname = outpath/'assemblies'\n",
    "assemblies_dirname.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "14bf8ffe-c818-44c2-9c55-283fd9d76e83",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in (outpath/'denovo').iterdir():\n",
    "    src = i/'contigs.fasta'\n",
    "    dst = assemblies_dirname/(i.name + '.fasta')\n",
    "    shutil.copyfile(src, dst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2c18f7a7-086c-459d-8364-638147f53039",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_dirname = outpath/'cluster'\n",
    "cluster_dirname.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c738b60-e912-401f-a651-f0a293ade9e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "threads = 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b81ede2-9418-499f-9f28-48fa8d327948",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler clustering\u001b[0m \u001b[2m(2022-01-27 14:44:15)\u001b[0m\n",
      "\u001b[2m    Trycycler cluster is a tool for clustering the contigs from multiple\u001b[0m\n",
      "\u001b[2mdifferent assemblies (e.g. from different assemblers) into highly-similar\u001b[0m\n",
      "\u001b[2mgroups.\u001b[0m\n",
      "\n",
      "Input assemblies:\n",
      "  A: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_01.fasta (4,975,261 bp, 1 contig)\n",
      "  B: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_02.fasta (4,974,173 bp, 1 contig)\n",
      "  C: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_03.fasta (4,974,706 bp, 1 contig)\n",
      "  D: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_04.fasta (4,975,117 bp, 1 contig)\n",
      "  E: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_05.fasta (4,974,073 bp, 1 contig)\n",
      "  F: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_06.fasta (4,975,311 bp, 1 contig)\n",
      "  G: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_07.fasta (4,974,101 bp, 1 contig)\n",
      "  H: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_08.fasta (4,975,257 bp, 1 contig)\n",
      "  I: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_09.fasta (4,975,052 bp, 1 contig)\n",
      "  J: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_10.fasta (4,975,066 bp, 1 contig)\n",
      "  K: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_11.fasta (4,974,546 bp, 1 contig)\n",
      "  L: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_12.fasta (4,975,169 bp, 1 contig)\n",
      "\n",
      "Output directory already exists: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster\n",
      "\n",
      "Input reads: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/barcode03.fastq.gz\n",
      "  50,065 reads (586,284,896 bp)\n",
      "  N50 = 20,749 bp\n",
      "\n",
      "Checking required software:\n",
      "      Mash: v2.3\n",
      "         R: v4.1.1\n",
      "       ape: v5.5\n",
      "  phangorn: v2.7.1\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mGetting contig depths\u001b[0m \u001b[2m(2022-01-27 14:44:28)\u001b[0m\n",
      "\u001b[2m    Trycycler now aligns the reads to each of the assemblies to assign a read\u001b[0m\n",
      "\u001b[2mdepth value to each of the contigs. Contigs displayed in red have a low read\u001b[0m\n",
      "\u001b[2mdepth and will be filtered out.\u001b[0m\n",
      "\n",
      "A (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_01.fasta):\n",
      "  50,059 alignments, mean depth = 118.5x\n",
      "  A_contig_1_polish: 4,975,261 bp, 118.5x\n",
      "\n",
      "B (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_02.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  B_contig_1_polish: 4,974,173 bp, 118.4x\n",
      "\n",
      "C (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_03.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  C_contig_1_polish: 4,974,706 bp, 118.4x\n",
      "\n",
      "D (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_04.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  D_contig_1_polish: 4,975,117 bp, 118.4x\n",
      "\n",
      "E (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_05.fasta):\n",
      "  50,060 alignments, mean depth = 118.5x\n",
      "  E_contig_1_polish: 4,974,073 bp, 118.5x\n",
      "\n",
      "F (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_06.fasta):\n",
      "  50,060 alignments, mean depth = 118.5x\n",
      "  F_contig_1_polish: 4,975,311 bp, 118.5x\n",
      "\n",
      "G (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_07.fasta):\n",
      "  50,060 alignments, mean depth = 118.5x\n",
      "  G_contig_1_polish: 4,974,101 bp, 118.5x\n",
      "\n",
      "H (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_08.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  H_contig_1_polish: 4,975,257 bp, 118.4x\n",
      "\n",
      "I (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_09.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  I_contig_1_polish: 4,975,052 bp, 118.4x\n",
      "\n",
      "J (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_10.fasta):\n",
      "  50,060 alignments, mean depth = 118.4x\n",
      "  J_contig_1_polish: 4,975,066 bp, 118.4x\n",
      "\n",
      "K (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_11.fasta):\n",
      "  50,059 alignments, mean depth = 118.5x\n",
      "  K_contig_1_polish: 4,974,546 bp, 118.5x\n",
      "\n",
      "L (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_12.fasta):\n",
      "  50,060 alignments, mean depth = 118.5x\n",
      "  L_contig_1_polish: 4,975,169 bp, 118.5x\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFiltering contigs\u001b[0m \u001b[2m(2022-01-27 14:49:31)\u001b[0m\n",
      "\u001b[2m    Contigs are now filtered out if they are too short (<1,000 bp) or too low\u001b[0m\n",
      "\u001b[2mdepth (<0.1 times the mean depth).\u001b[0m\n",
      "\n",
      "A (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_01.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "B (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_02.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "C (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_03.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "D (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_04.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "E (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_05.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "F (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_06.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "G (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_07.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "H (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_08.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "I (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_09.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "J (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_10.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "K (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_11.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "L (/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/assemblies/sample_12.fasta):\n",
      "  all contigs passed filtering\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mBuilding distance matrix\u001b[0m \u001b[2m(2022-01-27 14:49:31)\u001b[0m\n",
      "\u001b[2m    Mash is used to build a distance matrix of all contigs in the assemblies.\u001b[0m\n",
      "\n",
      "A_contig_1_polish: \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "B_contig_1_polish: 0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "C_contig_1_polish: 0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "D_contig_1_polish: 0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "E_contig_1_polish: 0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "F_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "G_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000\n",
      "H_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000\n",
      "I_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000\n",
      "J_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000\n",
      "K_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000\n",
      "L_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mClustering\u001b[0m \u001b[2m(2022-01-27 14:50:09)\u001b[0m\n",
      "\u001b[2m    The contigs are now split into clusters using a complete-linkage\u001b[0m\n",
      "\u001b[2mhierarchical approach.\u001b[0m\n",
      "\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/A_contig_1_polish.fasta: 4,975,261 bp, 118.5x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/B_contig_1_polish.fasta: 4,974,173 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/C_contig_1_polish.fasta: 4,974,706 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/D_contig_1_polish.fasta: 4,975,117 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/E_contig_1_polish.fasta: 4,974,073 bp, 118.5x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/F_contig_1_polish.fasta: 4,975,311 bp, 118.5x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/G_contig_1_polish.fasta: 4,974,101 bp, 118.5x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/H_contig_1_polish.fasta: 4,975,257 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/I_contig_1_polish.fasta: 4,975,052 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/J_contig_1_polish.fasta: 4,975,066 bp, 118.4x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/K_contig_1_polish.fasta: 4,974,546 bp, 118.5x\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/cluster_001/1_contigs/L_contig_1_polish.fasta: 4,975,169 bp, 118.5x\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mBuilding FastME tree\u001b[0m \u001b[2m(2022-01-27 14:50:09)\u001b[0m\n",
      "\u001b[2m    R (ape and phangorn) are used to build a FastME tree of the relationships\u001b[0m\n",
      "\u001b[2mbetween the contigs.\u001b[0m\n",
      "\n",
      "saving distance matrix: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/contigs.phylip\n",
      "saving tree: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode11/cluster/contigs.newick\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinished!\u001b[0m \u001b[2m(2022-01-27 14:50:11)\u001b[0m\n",
      "\u001b[2m    Now you must decide which clusters are good (i.e. contain well-assembled\u001b[0m\n",
      "\u001b[2mcontigs for replicons in the genome) and which are bad (i.e. contain incomplete\u001b[0m\n",
      "\u001b[2mor spurious contigs). You can then delete the directories corresponding to the\u001b[0m\n",
      "\u001b[2mbad clusters and proceed to the next step in the pipeline: trycycler reconcile.\u001b[0m\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler cluster -r {LR} -a {assemblies_dirname/'*'} -o {cluster_dirname} -t {threads}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "86179b2a-cd06-4651-bc0d-d06ba0dc3823",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler reconcile\u001b[0m \u001b[2m(2022-01-27 14:51:36)\u001b[0m\n",
      "\u001b[2m    Trycycler reconcile is a tool for reconciling multiple alternative contigs\u001b[0m\n",
      "\u001b[2mwith each other.\u001b[0m\n",
      "\n",
      "Input reads: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/barcode03.fastq.gz\n",
      "  size = 612,157,240 bytes\n",
      "\n",
      "Input contigs:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/A_contig_1_polish.fasta (4,975,261 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/B_contig_1_polish.fasta (4,974,173 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/C_contig_1_polish.fasta (4,974,706 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/D_contig_1_polish.fasta (4,975,117 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/E_contig_1_polish.fasta (4,974,073 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/F_contig_1_polish.fasta (4,975,311 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/G_contig_1_polish.fasta (4,974,101 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/H_contig_1_polish.fasta (4,975,257 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/I_contig_1_polish.fasta (4,975,052 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/J_contig_1_polish.fasta (4,975,066 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/K_contig_1_polish.fasta (4,974,546 bp)\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/1_contigs/L_contig_1_polish.fasta (4,975,169 bp)\n",
      "\n",
      "Checking required software:\n",
      "  minimap2: v2.22-r1101\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mInitial check of contigs\u001b[0m \u001b[2m(2022-01-27 14:51:37)\u001b[0m\n",
      "\u001b[2m    Before proceeding, Trycycler ensures that the input contigs appear\u001b[0m\n",
      "\u001b[2msufficiently close to each other to make a consensus. If not, the program will\u001b[0m\n",
      "\u001b[2mquit and the user must fix the input contigs (make them more similar to each\u001b[0m\n",
      "\u001b[2mother) or exclude some before trying again.\u001b[0m\n",
      "\n",
      "Relative sequence lengths:\n",
      "  A_contig_1_polish: \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  B_contig_1_polish: 1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  C_contig_1_polish: 1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  D_contig_1_polish: 1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  E_contig_1_polish: 1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  F_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000  1.000\n",
      "  G_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000  1.000\n",
      "  H_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000  1.000\n",
      "  I_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000  1.000\n",
      "  J_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000  1.000\n",
      "  K_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m  1.000\n",
      "  L_contig_1_polish: 1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  1.000  \u001b[2m1.000\u001b[0m\n",
      "\n",
      "Mash distances:\n",
      "  A_contig_1_polish: \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  B_contig_1_polish: 0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  C_contig_1_polish: 0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  D_contig_1_polish: 0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  E_contig_1_polish: 0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  F_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000  0.000\n",
      "  G_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000  0.000\n",
      "  H_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000  0.000\n",
      "  I_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000  0.000\n",
      "  J_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000  0.000\n",
      "  K_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m  0.000\n",
      "  L_contig_1_polish: 0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  0.000  \u001b[2m0.000\u001b[0m\n",
      "\n",
      "Contigs have passed the initial check - they seem sufficiently close to reconcile.\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mNormalising strands\u001b[0m \u001b[2m(2022-01-27 14:52:04)\u001b[0m\n",
      "\u001b[2m    In this step, Trycycler ensures that all sequences are on the same strand.\u001b[0m\n",
      "\u001b[2mIt does this by first finding a sequence that occurs once in each contig and\u001b[0m\n",
      "\u001b[2mthen flipping any of the contigs (converting to their reverse complement\u001b[0m\n",
      "\u001b[2msequence) which have this sequence on the negative strand.\u001b[0m\n",
      "\n",
      "Randomly-chosen common sequence:\n",
      "  TGAGTTCCTTTGGAAACTCGCTGGCCGTGAAGAATATGAAGGTTTTGATC\n",
      "  CATCTAAATCGACGGCCGCAGCGCGCGGACGTAACTTTGCTCAGGTAAGT\n",
      "  AACCTGCCGGTGGTGCTAATTGAAGGTGATCGCACCACGGACAACGCGAA\n",
      "  ACAGCGTGCTTTTGACTGGGATGAACTGAAATCACTGTATAATGGTCGGG\n",
      "  CATCCCGCGCAGTAGGCATCAAATCCAACAATAACGAGACATATGAACCG\n",
      "\n",
      "A_contig_1_polish: + strand (using original sequence)\n",
      "B_contig_1_polish: - strand (using reverse complement)\n",
      "C_contig_1_polish: + strand (using original sequence)\n",
      "D_contig_1_polish: + strand (using original sequence)\n",
      "E_contig_1_polish: + strand (using original sequence)\n",
      "F_contig_1_polish: + strand (using original sequence)\n",
      "G_contig_1_polish: - strand (using reverse complement)\n",
      "H_contig_1_polish: + strand (using original sequence)\n",
      "I_contig_1_polish: + strand (using original sequence)\n",
      "J_contig_1_polish: + strand (using original sequence)\n",
      "K_contig_1_polish: - strand (using reverse complement)\n",
      "L_contig_1_polish: + strand (using original sequence)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mCircularisation\u001b[0m \u001b[2m(2022-01-27 14:52:15)\u001b[0m\n",
      "\u001b[2m    Trycycler now compares the contigs to each other to repair any\u001b[0m\n",
      "\u001b[2mcircularisation issues. After this step, each sequence should be cleanly\u001b[0m\n",
      "\u001b[2mcircularised - i.e. the first base in the contig immediately follows the last\u001b[0m\n",
      "\u001b[2mbase. Each contig will be circularised by looking for the position of its start\u001b[0m\n",
      "\u001b[2mand end in the other contigs. If necessary, additional sequence will be added\u001b[0m\n",
      "\u001b[2mor duplicated sequence will be removed. If there are multiple possible ways to\u001b[0m\n",
      "\u001b[2mfix a contig's circularisation, then Trycycler will use read alignments to\u001b[0m\n",
      "\u001b[2mchoose the best one.\u001b[0m\n",
      "\n",
      "Circularising A_contig_1_polish:\n",
      "  using B_contig_1_polish:\n",
      "    circularising A_contig_1_polish by adding 40 bp of sequence from B_contig_1_polish (4365505-4365545)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in C_contig_1_polish\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in D_contig_1_polish\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in E_contig_1_polish\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in F_contig_1_polish\n",
      "  using G_contig_1_polish:\n",
      "    circularising A_contig_1_polish by adding 40 bp of sequence from G_contig_1_polish (4365265-4365305)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in H_contig_1_polish\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in I_contig_1_polish\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: A_contig_1_polish's end could not be found in J_contig_1_polish\n",
      "  using K_contig_1_polish:\n",
      "    circularising A_contig_1_polish by adding 40 bp of sequence from K_contig_1_polish (4365765-4365805)\n",
      "  using L_contig_1_polish:\n",
      "    circularising A_contig_1_polish by adding 40 bp of sequence from L_contig_1_polish (657725-657765)\n",
      "  circularisation complete (4,975,301 bp)\n",
      "\n",
      "Circularising B_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from A_contig_1_polish (608784-608826)\n",
      "  using C_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from C_contig_1_polish (608663-608705)\n",
      "  using D_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from D_contig_1_polish (608677-608719)\n",
      "  using E_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from E_contig_1_polish (608795-608837)\n",
      "  using F_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from F_contig_1_polish (608855-608897)\n",
      "  using G_contig_1_polish:\n",
      "    unable to circularise: cannot determine trim amount\n",
      "  using H_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from H_contig_1_polish (608838-608880)\n",
      "  using I_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from I_contig_1_polish (608645-608687)\n",
      "  using J_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from J_contig_1_polish (608645-608687)\n",
      "  using K_contig_1_polish:\n",
      "    unable to circularise: B_contig_1_polish's start/end is the same as K_contig_1_polish's start/end\n",
      "  using L_contig_1_polish:\n",
      "    circularising B_contig_1_polish by adding 42 bp of sequence from L_contig_1_polish (1266488-1266530)\n",
      "  circularisation complete (4,974,215 bp)\n",
      "\n",
      "Circularising C_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start could not be found in A_contig_1_polish\n",
      "  using B_contig_1_polish:\n",
      "    circularising C_contig_1_polish by adding 47 bp of sequence from B_contig_1_polish (4365442-4365489)\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising C_contig_1_polish by adding 47 bp of sequence from G_contig_1_polish (4365202-4365249)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: C_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising C_contig_1_polish by adding 48 bp of sequence from K_contig_1_polish (4365701-4365749)\n",
      "  using L_contig_1_polish:\n",
      "    circularising C_contig_1_polish by adding 48 bp of sequence from L_contig_1_polish (657661-657709)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,974,753 bp): score = 847,206,202\n",
      "    alternative 2 (4,974,754 bp): score = 847,206,170\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,974,753 bp)\n",
      "\n",
      "Circularising D_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start could not be found in A_contig_1_polish\n",
      "  using B_contig_1_polish:\n",
      "    circularising D_contig_1_polish by adding 47 bp of sequence from B_contig_1_polish (4365442-4365489)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising D_contig_1_polish by adding 47 bp of sequence from G_contig_1_polish (4365202-4365249)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: D_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising D_contig_1_polish by adding 48 bp of sequence from K_contig_1_polish (4365701-4365749)\n",
      "  using L_contig_1_polish:\n",
      "    circularising D_contig_1_polish by adding 48 bp of sequence from L_contig_1_polish (657661-657709)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,975,164 bp): score = 847,135,816\n",
      "    alternative 2 (4,975,165 bp): score = 847,135,782\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,975,164 bp)\n",
      "\n",
      "Circularising E_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start could not be found in A_contig_1_polish\n",
      "  using B_contig_1_polish:\n",
      "    circularising E_contig_1_polish by adding 47 bp of sequence from B_contig_1_polish (4365442-4365489)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising E_contig_1_polish by adding 47 bp of sequence from G_contig_1_polish (4365202-4365249)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: E_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising E_contig_1_polish by adding 48 bp of sequence from K_contig_1_polish (4365701-4365749)\n",
      "  using L_contig_1_polish:\n",
      "    circularising E_contig_1_polish by adding 48 bp of sequence from L_contig_1_polish (657661-657709)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,974,120 bp): score = 847,655,043\n",
      "    alternative 2 (4,974,121 bp): score = 847,655,011\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,974,120 bp)\n",
      "\n",
      "Circularising F_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start could not be found in A_contig_1_polish\n",
      "  using B_contig_1_polish:\n",
      "    circularising F_contig_1_polish by adding 52 bp of sequence from B_contig_1_polish (4365442-4365494)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising F_contig_1_polish by adding 52 bp of sequence from G_contig_1_polish (4365202-4365254)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: F_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising F_contig_1_polish by adding 53 bp of sequence from K_contig_1_polish (4365701-4365754)\n",
      "  using L_contig_1_polish:\n",
      "    circularising F_contig_1_polish by adding 53 bp of sequence from L_contig_1_polish (657661-657714)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,975,363 bp): score = 847,548,831\n",
      "    alternative 2 (4,975,364 bp): score = 847,548,795\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,975,363 bp)\n",
      "\n",
      "Circularising G_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from A_contig_1_polish (608813-608817)\n",
      "  using B_contig_1_polish:\n",
      "    unable to circularise: G_contig_1_polish's start/end is the same as B_contig_1_polish's start/end\n",
      "  using C_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from C_contig_1_polish (608692-608696)\n",
      "  using D_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from D_contig_1_polish (608706-608710)\n",
      "  using E_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from E_contig_1_polish (608824-608828)\n",
      "  using F_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from F_contig_1_polish (608884-608888)\n",
      "  using H_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from H_contig_1_polish (608867-608871)\n",
      "  using I_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from I_contig_1_polish (608674-608678)\n",
      "  using J_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from J_contig_1_polish (608674-608678)\n",
      "  using K_contig_1_polish:\n",
      "    unable to circularise: G_contig_1_polish's start/end is the same as K_contig_1_polish's start/end\n",
      "  using L_contig_1_polish:\n",
      "    circularising G_contig_1_polish by adding 4 bp of sequence from L_contig_1_polish (1266517-1266521)\n",
      "  circularisation complete (4,974,105 bp)\n",
      "\n",
      "Circularising H_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start could not be found in A_contig_1_polish\n",
      "  using B_contig_1_polish:\n",
      "    circularising H_contig_1_polish by adding 47 bp of sequence from B_contig_1_polish (4365442-4365489)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising H_contig_1_polish by adding 47 bp of sequence from G_contig_1_polish (4365202-4365249)\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: H_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising H_contig_1_polish by adding 48 bp of sequence from K_contig_1_polish (4365701-4365749)\n",
      "  using L_contig_1_polish:\n",
      "    circularising H_contig_1_polish by adding 48 bp of sequence from L_contig_1_polish (657661-657709)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,975,304 bp): score = 847,263,857\n",
      "    alternative 2 (4,975,305 bp): score = 847,263,825\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,975,304 bp)\n",
      "\n",
      "Circularising I_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as A_contig_1_polish's start/end\n",
      "  using B_contig_1_polish:\n",
      "    circularising I_contig_1_polish by adding 60 bp of sequence from B_contig_1_polish (4365442-4365502)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising I_contig_1_polish by adding 60 bp of sequence from G_contig_1_polish (4365202-4365262)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using J_contig_1_polish:\n",
      "    unable to circularise: I_contig_1_polish's start/end is the same as J_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising I_contig_1_polish by adding 61 bp of sequence from K_contig_1_polish (4365701-4365762)\n",
      "  using L_contig_1_polish:\n",
      "    circularising I_contig_1_polish by adding 61 bp of sequence from L_contig_1_polish (657661-657722)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,975,112 bp): score = 847,255,215\n",
      "    alternative 2 (4,975,113 bp): score = 847,255,179\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,975,112 bp)\n",
      "\n",
      "Circularising J_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as A_contig_1_polish's start/end\n",
      "  using B_contig_1_polish:\n",
      "    circularising J_contig_1_polish by adding 60 bp of sequence from B_contig_1_polish (4365442-4365502)\n",
      "  using C_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as C_contig_1_polish's start/end\n",
      "  using D_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as D_contig_1_polish's start/end\n",
      "  using E_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as E_contig_1_polish's start/end\n",
      "  using F_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as F_contig_1_polish's start/end\n",
      "  using G_contig_1_polish:\n",
      "    circularising J_contig_1_polish by adding 60 bp of sequence from G_contig_1_polish (4365202-4365262)\n",
      "  using H_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as H_contig_1_polish's start/end\n",
      "  using I_contig_1_polish:\n",
      "    unable to circularise: J_contig_1_polish's start/end is the same as I_contig_1_polish's start/end\n",
      "  using K_contig_1_polish:\n",
      "    circularising J_contig_1_polish by adding 61 bp of sequence from K_contig_1_polish (4365701-4365762)\n",
      "  using L_contig_1_polish:\n",
      "    circularising J_contig_1_polish by adding 61 bp of sequence from L_contig_1_polish (657661-657722)\n",
      "  choosing best circularisation of 2 alternatives\n",
      "    alternative 1 (4,975,126 bp): score = 847,252,069\n",
      "    alternative 2 (4,975,127 bp): score = 847,252,033\n",
      "    best alternative: 1\n",
      "  circularisation complete (4,975,126 bp)\n",
      "\n",
      "Circularising K_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from A_contig_1_polish (608764-608817)\n",
      "  using B_contig_1_polish:\n",
      "    unable to circularise: K_contig_1_polish's start/end is the same as B_contig_1_polish's start/end\n",
      "  using C_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from C_contig_1_polish (608643-608696)\n",
      "  using D_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from D_contig_1_polish (608657-608710)\n",
      "  using E_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from E_contig_1_polish (608775-608828)\n",
      "  using F_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from F_contig_1_polish (608835-608888)\n",
      "  using G_contig_1_polish:\n",
      "    unable to circularise: K_contig_1_polish's start/end is the same as G_contig_1_polish's start/end\n",
      "  using H_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from H_contig_1_polish (608818-608871)\n",
      "  using I_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from I_contig_1_polish (608625-608678)\n",
      "  using J_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from J_contig_1_polish (608625-608678)\n",
      "  using L_contig_1_polish:\n",
      "    circularising K_contig_1_polish by adding 53 bp of sequence from L_contig_1_polish (1266468-1266521)\n",
      "  circularisation complete (4,974,599 bp)\n",
      "\n",
      "Circularising L_contig_1_polish:\n",
      "  using A_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from A_contig_1_polish (4317475-4317536)\n",
      "  using B_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from B_contig_1_polish (3707758-3707819)\n",
      "  using C_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from C_contig_1_polish (4317008-4317069)\n",
      "  using D_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from D_contig_1_polish (4317387-4317448)\n",
      "  using E_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from E_contig_1_polish (4316458-4316519)\n",
      "  using F_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from F_contig_1_polish (4317581-4317642)\n",
      "  using G_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from G_contig_1_polish (3707598-3707659)\n",
      "  using H_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from H_contig_1_polish (4317526-4317587)\n",
      "  using I_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from I_contig_1_polish (4317325-4317386)\n",
      "  using J_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from J_contig_1_polish (4317337-4317398)\n",
      "  using K_contig_1_polish:\n",
      "    circularising L_contig_1_polish by adding 61 bp of sequence from K_contig_1_polish (3708031-3708092)\n",
      "  circularisation complete (4,975,230 bp)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinding starting sequence\u001b[0m \u001b[2m(2022-01-27 14:59:33)\u001b[0m\n",
      "\u001b[2m    In this step, Trycycler finds a sequence to use as a starting point for\u001b[0m\n",
      "\u001b[2meach of the contigs. This can be a standard starting point (e.g. the dnaA gene)\u001b[0m\n",
      "\u001b[2mor if one is not found, then a randomly-chosen unique sequence will be used. If\u001b[0m\n",
      "\u001b[2mnecessary, the sequences will be flipped (converted to their reverse complement\u001b[0m\n",
      "\u001b[2msequence) to ensure that the starting sequence is on the positive strand.\u001b[0m\n",
      "\n",
      "Looking for known starting sequences in each contig...\n",
      "\n",
      "Found starting sequence 0000_A1D48_RS00005 (chromosomal replication initiator protein DnaA)\n",
      "  GTGTCACTTTCGCTTTGGCAGCAGTGTCTTGCCCGATTGCAGGATGAGTT...\n",
      "\n",
      "A_contig_1_polish: + strand (using original sequence)\n",
      "B_contig_1_polish: + strand (using original sequence)\n",
      "C_contig_1_polish: + strand (using original sequence)\n",
      "D_contig_1_polish: + strand (using original sequence)\n",
      "E_contig_1_polish: + strand (using original sequence)\n",
      "F_contig_1_polish: + strand (using original sequence)\n",
      "G_contig_1_polish: + strand (using original sequence)\n",
      "H_contig_1_polish: + strand (using original sequence)\n",
      "I_contig_1_polish: + strand (using original sequence)\n",
      "J_contig_1_polish: + strand (using original sequence)\n",
      "K_contig_1_polish: + strand (using original sequence)\n",
      "L_contig_1_polish: + strand (using original sequence)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mRotating contigs to starting sequence\u001b[0m \u001b[2m(2022-01-27 14:59:44)\u001b[0m\n",
      "\u001b[2m    For a circular contig, any point in the sequence is a valid starting\u001b[0m\n",
      "\u001b[2mposition and it can thus be 'rotated' by moving sequence from the contig start\u001b[0m\n",
      "\u001b[2mto the contig end. In this step, Trycycler rotates each contig such that it\u001b[0m\n",
      "\u001b[2mbegins with the starting sequence, ensuring that all contigs begin and end\u001b[0m\n",
      "\u001b[2mtogether so they can be aligned to each other.\u001b[0m\n",
      "\n",
      "A_contig_1_polish: rotating by 4,294,844 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,301 bp)\n",
      "\n",
      "B_contig_1_polish: rotating by 3,685,140 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,974,215 bp)\n",
      "\n",
      "C_contig_1_polish: rotating by 4,294,390 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,974,753 bp)\n",
      "\n",
      "D_contig_1_polish: rotating by 4,294,756 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,164 bp)\n",
      "\n",
      "E_contig_1_polish: rotating by 4,293,835 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,974,120 bp)\n",
      "\n",
      "F_contig_1_polish: rotating by 4,294,950 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,363 bp)\n",
      "\n",
      "G_contig_1_polish: rotating by 3,684,969 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,974,105 bp)\n",
      "\n",
      "H_contig_1_polish: rotating by 4,294,895 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,304 bp)\n",
      "\n",
      "I_contig_1_polish: rotating by 4,294,694 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,112 bp)\n",
      "\n",
      "J_contig_1_polish: rotating by 4,294,706 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,126 bp)\n",
      "\n",
      "K_contig_1_polish: rotating by 3,685,406 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,974,599 bp)\n",
      "\n",
      "L_contig_1_polish: rotating by 4,952,538 bp\n",
      "   GTGTCACTTTCGCTTTGGCA...TTGTTCGAGTGGAGTCCGCC (4,975,230 bp)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mPairwise global alignments\u001b[0m \u001b[2m(2022-01-27 14:59:49)\u001b[0m\n",
      "\u001b[2m    Trycycler uses the edlib aligner to get global alignments between all pairs\u001b[0m\n",
      "\u001b[2mof sequences. This can help you to spot any problematic sequences that should\u001b[0m\n",
      "\u001b[2mbe excluded before continuing. If you see any sequences with notably worse\u001b[0m\n",
      "\u001b[2midentities or max indels, you can remove them (delete the contig's FASTA) and\u001b[0m\n",
      "\u001b[2mrun this command again.\u001b[0m\n",
      "\n",
      "A_contig_1_polish vs B_contig_1_polish... 99.95% identity, max indel = 9\n",
      "A_contig_1_polish vs C_contig_1_polish... 99.97% identity, max indel = 11\n",
      "A_contig_1_polish vs D_contig_1_polish... 99.98% identity, max indel = 11\n",
      "A_contig_1_polish vs E_contig_1_polish... 99.94% identity, max indel = 6\n",
      "A_contig_1_polish vs F_contig_1_polish... 99.98% identity, max indel = 7\n",
      "A_contig_1_polish vs G_contig_1_polish... 99.94% identity, max indel = 8\n",
      "A_contig_1_polish vs H_contig_1_polish... 99.98% identity, max indel = 11\n",
      "A_contig_1_polish vs I_contig_1_polish... 99.98% identity, max indel = 9\n",
      "A_contig_1_polish vs J_contig_1_polish... 99.98% identity, max indel = 9\n",
      "A_contig_1_polish vs K_contig_1_polish... 99.95% identity, max indel = 7\n",
      "A_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 6\n",
      "B_contig_1_polish vs C_contig_1_polish... 99.94% identity, max indel = 5\n",
      "B_contig_1_polish vs D_contig_1_polish... 99.94% identity, max indel = 5\n",
      "B_contig_1_polish vs E_contig_1_polish... 99.94% identity, max indel = 6\n",
      "B_contig_1_polish vs F_contig_1_polish... 99.95% identity, max indel = 11\n",
      "B_contig_1_polish vs G_contig_1_polish... 99.93% identity, max indel = 7\n",
      "B_contig_1_polish vs H_contig_1_polish... 99.94% identity, max indel = 7\n",
      "B_contig_1_polish vs I_contig_1_polish... 99.96% identity, max indel = 6\n",
      "B_contig_1_polish vs J_contig_1_polish... 99.96% identity, max indel = 5\n",
      "B_contig_1_polish vs K_contig_1_polish... 99.94% identity, max indel = 13\n",
      "B_contig_1_polish vs L_contig_1_polish... 99.95% identity, max indel = 6\n",
      "C_contig_1_polish vs D_contig_1_polish... 99.98% identity, max indel = 4\n",
      "C_contig_1_polish vs E_contig_1_polish... 99.94% identity, max indel = 7\n",
      "C_contig_1_polish vs F_contig_1_polish... 99.97% identity, max indel = 5\n",
      "C_contig_1_polish vs G_contig_1_polish... 99.94% identity, max indel = 9\n",
      "C_contig_1_polish vs H_contig_1_polish... 99.97% identity, max indel = 6\n",
      "C_contig_1_polish vs I_contig_1_polish... 99.96% identity, max indel = 6\n",
      "C_contig_1_polish vs J_contig_1_polish... 99.96% identity, max indel = 8\n",
      "C_contig_1_polish vs K_contig_1_polish... 99.96% identity, max indel = 8\n",
      "C_contig_1_polish vs L_contig_1_polish... 99.97% identity, max indel = 6\n",
      "D_contig_1_polish vs E_contig_1_polish... 99.94% identity, max indel = 7\n",
      "D_contig_1_polish vs F_contig_1_polish... 99.98% identity, max indel = 5\n",
      "D_contig_1_polish vs G_contig_1_polish... 99.95% identity, max indel = 9\n",
      "D_contig_1_polish vs H_contig_1_polish... 99.98% identity, max indel = 6\n",
      "D_contig_1_polish vs I_contig_1_polish... 99.97% identity, max indel = 6\n",
      "D_contig_1_polish vs J_contig_1_polish... 99.97% identity, max indel = 6\n",
      "D_contig_1_polish vs K_contig_1_polish... 99.96% identity, max indel = 8\n",
      "D_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 6\n",
      "E_contig_1_polish vs F_contig_1_polish... 99.94% identity, max indel = 11\n",
      "E_contig_1_polish vs G_contig_1_polish... 99.94% identity, max indel = 5\n",
      "E_contig_1_polish vs H_contig_1_polish... 99.94% identity, max indel = 6\n",
      "E_contig_1_polish vs I_contig_1_polish... 99.95% identity, max indel = 6\n",
      "E_contig_1_polish vs J_contig_1_polish... 99.94% identity, max indel = 6\n",
      "E_contig_1_polish vs K_contig_1_polish... 99.93% identity, max indel = 7\n",
      "E_contig_1_polish vs L_contig_1_polish... 99.94% identity, max indel = 5\n",
      "F_contig_1_polish vs G_contig_1_polish... 99.94% identity, max indel = 7\n",
      "F_contig_1_polish vs H_contig_1_polish... 99.98% identity, max indel = 8\n",
      "F_contig_1_polish vs I_contig_1_polish... 99.98% identity, max indel = 11\n",
      "F_contig_1_polish vs J_contig_1_polish... 99.98% identity, max indel = 11\n",
      "F_contig_1_polish vs K_contig_1_polish... 99.95% identity, max indel = 8\n",
      "F_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 7\n",
      "G_contig_1_polish vs H_contig_1_polish... 99.94% identity, max indel = 9\n",
      "G_contig_1_polish vs I_contig_1_polish... 99.94% identity, max indel = 8\n",
      "G_contig_1_polish vs J_contig_1_polish... 99.94% identity, max indel = 8\n",
      "G_contig_1_polish vs K_contig_1_polish... 99.94% identity, max indel = 9\n",
      "G_contig_1_polish vs L_contig_1_polish... 99.94% identity, max indel = 5\n",
      "H_contig_1_polish vs I_contig_1_polish... 99.98% identity, max indel = 6\n",
      "H_contig_1_polish vs J_contig_1_polish... 99.98% identity, max indel = 9\n",
      "H_contig_1_polish vs K_contig_1_polish... 99.96% identity, max indel = 8\n",
      "H_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 9\n",
      "I_contig_1_polish vs J_contig_1_polish... 99.99% identity, max indel = 4\n",
      "I_contig_1_polish vs K_contig_1_polish... 99.95% identity, max indel = 13\n",
      "I_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 8\n",
      "J_contig_1_polish vs K_contig_1_polish... 99.96% identity, max indel = 13\n",
      "J_contig_1_polish vs L_contig_1_polish... 99.98% identity, max indel = 8\n",
      "K_contig_1_polish vs L_contig_1_polish... 99.95% identity, max indel = 9\n",
      "\n",
      "Pairwise identities:\n",
      "  A_contig_1_polish: \u001b[2m100.00%\u001b[0m   99.95%   99.97%   99.98%   99.94%   99.98%   99.94%   99.98%   99.98%   99.98%   99.95%   99.98%\n",
      "  B_contig_1_polish:  99.95%  \u001b[2m100.00%\u001b[0m   99.94%   99.94%   99.94%   99.95%   99.93%   99.94%   99.96%   99.96%   99.94%   99.95%\n",
      "  C_contig_1_polish:  99.97%   99.94%  \u001b[2m100.00%\u001b[0m   99.98%   99.94%   99.97%   99.94%   99.97%   99.96%   99.96%   99.96%   99.97%\n",
      "  D_contig_1_polish:  99.98%   99.94%   99.98%  \u001b[2m100.00%\u001b[0m   99.94%   99.98%   99.95%   99.98%   99.97%   99.97%   99.96%   99.98%\n",
      "  E_contig_1_polish:  99.94%   99.94%   99.94%   99.94%  \u001b[2m100.00%\u001b[0m   99.94%   99.94%   99.94%   99.95%   99.94%   99.93%   99.94%\n",
      "  F_contig_1_polish:  99.98%   99.95%   99.97%   99.98%   99.94%  \u001b[2m100.00%\u001b[0m   99.94%   99.98%   99.98%   99.98%   99.95%   99.98%\n",
      "  G_contig_1_polish:  99.94%   99.93%   99.94%   99.95%   99.94%   99.94%  \u001b[2m100.00%\u001b[0m   99.94%   99.94%   99.94%   99.94%   99.94%\n",
      "  H_contig_1_polish:  99.98%   99.94%   99.97%   99.98%   99.94%   99.98%   99.94%  \u001b[2m100.00%\u001b[0m   99.98%   99.98%   99.96%   99.98%\n",
      "  I_contig_1_polish:  99.98%   99.96%   99.96%   99.97%   99.95%   99.98%   99.94%   99.98%  \u001b[2m100.00%\u001b[0m   99.99%   99.95%   99.98%\n",
      "  J_contig_1_polish:  99.98%   99.96%   99.96%   99.97%   99.94%   99.98%   99.94%   99.98%   99.99%  \u001b[2m100.00%\u001b[0m   99.96%   99.98%\n",
      "  K_contig_1_polish:  99.95%   99.94%   99.96%   99.96%   99.93%   99.95%   99.94%   99.96%   99.95%   99.96%  \u001b[2m100.00%\u001b[0m   99.95%\n",
      "  L_contig_1_polish:  99.98%   99.95%   99.97%   99.98%   99.94%   99.98%   99.94%   99.98%   99.98%   99.98%   99.95%  \u001b[2m100.00%\u001b[0m\n",
      "\n",
      "Maximum insertion/deletion sizes:\n",
      "  A_contig_1_polish: \u001b[2m 0\u001b[0m   9  11  11   6   7   8  11   9   9   7   6\n",
      "  B_contig_1_polish:  9  \u001b[2m 0\u001b[0m   5   5   6  11   7   7   6   5  13   6\n",
      "  C_contig_1_polish: 11   5  \u001b[2m 0\u001b[0m   4   7   5   9   6   6   8   8   6\n",
      "  D_contig_1_polish: 11   5   4  \u001b[2m 0\u001b[0m   7   5   9   6   6   6   8   6\n",
      "  E_contig_1_polish:  6   6   7   7  \u001b[2m 0\u001b[0m  11   5   6   6   6   7   5\n",
      "  F_contig_1_polish:  7  11   5   5  11  \u001b[2m 0\u001b[0m   7   8  11  11   8   7\n",
      "  G_contig_1_polish:  8   7   9   9   5   7  \u001b[2m 0\u001b[0m   9   8   8   9   5\n",
      "  H_contig_1_polish: 11   7   6   6   6   8   9  \u001b[2m 0\u001b[0m   6   9   8   9\n",
      "  I_contig_1_polish:  9   6   6   6   6  11   8   6  \u001b[2m 0\u001b[0m   4  13   8\n",
      "  J_contig_1_polish:  9   5   8   6   6  11   8   9   4  \u001b[2m 0\u001b[0m  13   8\n",
      "  K_contig_1_polish:  7  13   8   8   7   8   9   8  13  13  \u001b[2m 0\u001b[0m   9\n",
      "  L_contig_1_polish:  6   6   6   6   5   7   5   9   8   8   9  \u001b[2m 0\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinished!\u001b[0m \u001b[2m(2022-01-27 15:08:02)\u001b[0m\n",
      "\u001b[2m    All contig sequences are now reconciled and ready for the next step in the\u001b[0m\n",
      "\u001b[2mpipeline: trycycler msa.\u001b[0m\n",
      "\n",
      "Saving sequences to file: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/2_all_seqs.fasta\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler reconcile -c {cluster_dirname/'cluster_001'} -r {LR} -t {threads}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "08c1e097-0b1d-49eb-9aff-066148c4dfb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler MSA\u001b[0m \u001b[2m(2022-01-27 15:14:42)\u001b[0m\n",
      "\u001b[2m    Trycycler MSA is a tool for conducting global multiple sequence alignment\u001b[0m\n",
      "\u001b[2mof contig sequences.\u001b[0m\n",
      "\n",
      "Input sequences:\n",
      "  A_contig_1_polish: 4,975,301 bp\n",
      "  B_contig_1_polish: 4,974,215 bp\n",
      "  C_contig_1_polish: 4,974,753 bp\n",
      "  D_contig_1_polish: 4,975,164 bp\n",
      "  E_contig_1_polish: 4,974,120 bp\n",
      "  F_contig_1_polish: 4,975,363 bp\n",
      "  G_contig_1_polish: 4,974,105 bp\n",
      "  H_contig_1_polish: 4,975,304 bp\n",
      "  I_contig_1_polish: 4,975,112 bp\n",
      "  J_contig_1_polish: 4,975,126 bp\n",
      "  K_contig_1_polish: 4,974,599 bp\n",
      "  L_contig_1_polish: 4,975,230 bp\n",
      "\n",
      "Checking required software:\n",
      "   MUSCLE: v3.8.1551\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mPartitioning sequences\u001b[0m \u001b[2m(2022-01-27 15:14:42)\u001b[0m\n",
      "\u001b[2m    The sequences are now partitioned into smaller chunks to make the multiple\u001b[0m\n",
      "\u001b[2msequence alignment more tractable.\u001b[0m\n",
      "\n",
      "pieces: 4820\n",
      "\n",
      "median piece size: 1,000 bp\n",
      "max piece size:    3,000 bp\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mRunning Muscle\u001b[0m \u001b[2m(2022-01-27 15:16:29)\u001b[0m\n",
      "\u001b[2m    Trycycler now runs Muscle on each of the pieces to turn them into multiple\u001b[0m\n",
      "\u001b[2msequence alignments.\u001b[0m\n",
      "\n",
      "pieces: 4820\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mMerging MSA\u001b[0m \u001b[2m(2022-01-27 15:18:30)\u001b[0m\n",
      "\u001b[2m    Each of the MSA pieces are now merged together and saved to file.\u001b[0m\n",
      "\n",
      "MSA length: 4,977,368 bp\n",
      "Saving to: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/3_msa.fasta\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler msa -c {cluster_dirname/'cluster_001'} -t {threads}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9e6dd483-191f-41ea-a060-1096c1f4bdb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler partitioning\u001b[0m \u001b[2m(2022-01-27 15:18:56)\u001b[0m\n",
      "\u001b[2m    Trycycler partition is a tool for partitioning reads by cluster. I.e. each\u001b[0m\n",
      "\u001b[2mread will be assigned to one cluster and saved in a file for that cluster.\u001b[0m\n",
      "\n",
      "Input reads: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/fastq/barcode03.fastq.gz\n",
      "  50,065 reads (586,284,896 bp)\n",
      "  N50 = 20,749 bp\n",
      "\n",
      "Input clusters:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/2_all_seqs.fasta: 12 contigs, mean length = 4,974,866 bp\n",
      "\n",
      "Checking required software:\n",
      "  minimap2: v2.22-r1101\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mAligning reads to each contig\u001b[0m \u001b[2m(2022-01-27 15:19:08)\u001b[0m\n",
      "\u001b[2m    The reads are independently aligned to each of the contigs and Trycycler\u001b[0m\n",
      "\u001b[2mwill remember the single best alignment for each read.\u001b[0m\n",
      "\n",
      "A_contig_1_polish (4,975,301 bp): 50,660 alignments\n",
      "B_contig_1_polish (4,974,215 bp): 50,683 alignments\n",
      "C_contig_1_polish (4,974,753 bp): 50,679 alignments\n",
      "D_contig_1_polish (4,975,164 bp): 50,682 alignments\n",
      "E_contig_1_polish (4,974,120 bp): 50,668 alignments\n",
      "F_contig_1_polish (4,975,363 bp): 50,674 alignments\n",
      "G_contig_1_polish (4,974,105 bp): 50,666 alignments\n",
      "H_contig_1_polish (4,975,304 bp): 50,690 alignments\n",
      "I_contig_1_polish (4,975,112 bp): 50,697 alignments\n",
      "J_contig_1_polish (4,975,126 bp): 50,689 alignments\n",
      "K_contig_1_polish (4,974,599 bp): 50,667 alignments\n",
      "L_contig_1_polish (4,975,230 bp): 50,664 alignments\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mSaving reads to cluster directories\u001b[0m \u001b[2m(2022-01-27 15:23:03)\u001b[0m\n",
      "\u001b[2m    Reads are now saved in a file in the cluster directory to which they\u001b[0m\n",
      "\u001b[2maligned best.\u001b[0m\n",
      "\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/4_reads.fastq:\n",
      "  43,390 reads (86.67%)\n",
      "  574,319,550 bases (97.96%)\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler partition -r {LR} -c {cluster_dirname/'cluster_001'} -t {threads}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "47cef6d3-16b8-4239-8935-52fbfbf31bf2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mStarting Trycycler consensus\u001b[0m \u001b[2m(2022-01-27 15:23:19)\u001b[0m\n",
      "\u001b[2m    Trycycler consensus is the final stage of the Trycycler pipeline. It\u001b[0m\n",
      "\u001b[2moperates on one replicon (i.e. cluster) at a time. It takes the multiple\u001b[0m\n",
      "\u001b[2msequence alignment of alternative contig sequences and combines them into a\u001b[0m\n",
      "\u001b[2msingle consensus sequence. Where needed, it will use read alignments to help\u001b[0m\n",
      "\u001b[2mchoose which variants to include/exclude from the consensus sequence. If all\u001b[0m\n",
      "\u001b[2mgoes well, the final consensus will be free of any large-scale errors.\u001b[0m\n",
      "\n",
      "Input reads: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/4_reads.fastq\n",
      "  43,390 reads (574,319,550 bp)\n",
      "  N50 = 21,032 bp\n",
      "\n",
      "Input contigs: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/2_all_seqs.fasta\n",
      "  A_contig_1_polish: 4,975,301 bp\n",
      "  B_contig_1_polish: 4,974,215 bp\n",
      "  C_contig_1_polish: 4,974,753 bp\n",
      "  D_contig_1_polish: 4,975,164 bp\n",
      "  E_contig_1_polish: 4,974,120 bp\n",
      "  F_contig_1_polish: 4,975,363 bp\n",
      "  G_contig_1_polish: 4,974,105 bp\n",
      "  H_contig_1_polish: 4,975,304 bp\n",
      "  I_contig_1_polish: 4,975,112 bp\n",
      "  J_contig_1_polish: 4,975,126 bp\n",
      "  K_contig_1_polish: 4,974,599 bp\n",
      "  L_contig_1_polish: 4,975,230 bp\n",
      "\n",
      "Checking required software:\n",
      "  minimap2: v2.22-r1101\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mPartitioning MSA\u001b[0m \u001b[2m(2022-01-27 15:23:20)\u001b[0m\n",
      "\u001b[2m    The multiple sequence alignment is now partitioned into chunks. Chunks\u001b[0m\n",
      "\u001b[2mwhere the input contig sequences are all in agreement are called \"same\" chunks,\u001b[0m\n",
      "\u001b[2mand those where the input contig sequences disagree are called \"different\"\u001b[0m\n",
      "\u001b[2mchunks. The consensus sequence will be made by choosing a best option for each\u001b[0m\n",
      "\u001b[2mof the different chunks.\u001b[0m\n",
      "\n",
      "chunks: 10,997 (5,499 same, 5,498 different)\n",
      "combining small chunks: 9,135 (4,568 same, 4,567 different)\n",
      "\n",
      "Saving sequences to graph: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/5_chunked_sequence.gfa\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mInitial consensus\u001b[0m \u001b[2m(2022-01-27 15:23:37)\u001b[0m\n",
      "\u001b[2m    Trycycler now makes an initial consensus sequence by choosing a sequence\u001b[0m\n",
      "\u001b[2mfor each of the different chunks. The chosen sequence is the one with the\u001b[0m\n",
      "\u001b[2mlowest total Hamming distance to the other sequences. For example, a chunk with\u001b[0m\n",
      "\u001b[2moptions of TT, TT, CC, CC and TA will give a consensus of TT. If the total\u001b[0m\n",
      "\u001b[2mHamming distances fail to break a tie or if all sequences differ, the chunk\u001b[0m\n",
      "\u001b[2mwill be flagged for read-based assessment.\u001b[0m\n",
      "\n",
      "Consensus length: 4,975,182 bp\n",
      "\n",
      "Different chunks needing assessment:     50\n",
      "Different chunks not needing assessment: 4,517\n",
      "\n",
      "Saving sequence to file: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/6_initial_consensus.fasta\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mIndexing reads\u001b[0m \u001b[2m(2022-01-27 15:23:38)\u001b[0m\n",
      "\u001b[2m    Trycycler now aligns all reads to the initial consensus to form an index of\u001b[0m\n",
      "\u001b[2mwhich reads span each of the chunks. This makes the following step faster, as\u001b[0m\n",
      "\u001b[2monly relevant reads will be used when conducting read-based assessment of\u001b[0m\n",
      "\u001b[2mchunks.\u001b[0m\n",
      "\n",
      "Aligning reads to the initial consensus:\n",
      "  43,839 alignments\n",
      "\n",
      "Filtering for best alignment per read:\n",
      "  43,179 alignments\n",
      "\n",
      "Gathering reads for chunks: 50 / 50\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mChoosing best options with reads\u001b[0m \u001b[2m(2022-01-27 15:24:19)\u001b[0m\n",
      "\u001b[2m    For each of the chunks to be assessed, Trycycler now aligns the relevant\u001b[0m\n",
      "\u001b[2mreads to each alternative sequence. Whichever option gives the strongest read\u001b[0m\n",
      "\u001b[2malignments (defined as the total alignment score for each of the read's best\u001b[0m\n",
      "\u001b[2malignment) is chosen as the best. This should result in a consensus sequence\u001b[0m\n",
      "\u001b[2mwhich is more accurate than the initial consensus.\u001b[0m\n",
      "\n",
      "Processing chunks: 50 / 50\n",
      "\n",
      "Chunks where sequence is...\n",
      "  the same as in the initial consensus: 42\n",
      "  different to the initial consensus:   8\n",
      "\n",
      "Saving sequence to file: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/cluster/cluster_001/7_final_consensus.fasta\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run -n trycycler trycycler consensus -c {cluster_dirname/'cluster_001'} -t {threads}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "34e69094-b033-48af-9d11-aa4499850d92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "!cat {cluster_dirname/'*'/'7_final_consensus.fasta'} > {outpath/'final_consensus.fasta'}\n",
    "!grep '>' {outpath/'final_consensus.fasta'} | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "63e7bca6-36d7-4142-b6f1-d0277d64a54c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking to launch executable \"/usr/local/bin/bwa-mem2.avx2\", simd = .avx2\n",
      "Launching executable \"/usr/local/bin/bwa-mem2.avx2\"\n",
      "[bwa_index] Pack FASTA... 0.03 sec\n",
      "* Entering FMI_search\n",
      "init ticks = 380589866\n",
      "ref seq len = 9950284\n",
      "binary seq ticks = 194352432\n",
      "build suffix-array ticks = 2897193805\n",
      "ref_seq_len = 9950284\n",
      "count = 0, 2386454, 4975142, 7563830, 9950284\n",
      "BWT[7387327] = 4\n",
      "CP_SHIFT = 6, CP_MASK = 63\n",
      "sizeof CP_OCC = 64\n",
      "pos: 1243786, ref_seq_len__: 1243785\n",
      "max_occ_ind = 155473\n",
      "build fm-index ticks = 712721206\n",
      "Total time taken: 2.1531\n",
      "Looking to launch executable \"/usr/local/bin/bwa-mem2.avx2\", simd = .avx2\n",
      "Launching executable \"/usr/local/bin/bwa-mem2.avx2\"\n",
      "-----------------------------\n",
      "Executing in AVX2 mode!!\n",
      "-----------------------------\n",
      "* SA compression enabled with xfactor: 8\n",
      "* Ref file: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Entering FMI_search\n",
      "* Index file found. Loading index from /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta.bwt.2bit.64\n",
      "* Reference seq len for bi-index = 9950285\n",
      "* sentinel-index: 7387327\n",
      "* Count:\n",
      "0,\t1\n",
      "1,\t2386455\n",
      "2,\t4975143\n",
      "3,\t7563831\n",
      "4,\t9950285\n",
      "\n",
      "* Reading other elements of the index from files /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Index prefix: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Read 0 ALT contigs\n",
      "* Done reading Index!!\n",
      "* Reading reference genome..\n",
      "* Binary seq file = /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta.0123\n",
      "* Reference genome size: 9950284 bp\n",
      "* Done reading reference genome !!\n",
      "\n",
      "------------------------------------------\n",
      "1. Memory pre-allocation for Chaining: 4458.8284 MB\n",
      "2. Memory pre-allocation for BSW: 7667.7448 MB\n",
      "3. Memory pre-allocation for BWT: 2474.0536 MB\n",
      "------------------------------------------\n",
      "* Threads used (compute): 32\n",
      "* No. of pipeline threads: 2\n",
      "\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 88586528, nseq: 325457\n",
      "\t[0000][ M::kt_pipeline] read 325457 sequences (88586528 bp)...\n",
      "[0000] Calling mem_process_seqs.., task: 0\n",
      "[0000] 1. Calling kt_for - worker_bwt\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 0, nseq: 0\n",
      "[0000] 2. Calling kt_for - worker_aln\n",
      "[0000] 3. Calling kt_for - worker_sam\n",
      "\t[0000][ M::mem_process_seqs] Processed 325457 reads in 42.611 CPU sec, 1.450 real sec\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 0, nseq: 0\n",
      "[0000] Computation ends..\n",
      "No. of OMP threads: 32\n",
      "Processor is running @1995.501528 MHz\n",
      "Runtime profile:\n",
      "\n",
      "\tTime taken for main_mem function: 3.60 sec\n",
      "\n",
      "\tIO times (sec) :\n",
      "\tReading IO time (reads) avg: 1.63, (1.63, 1.63)\n",
      "\tWriting IO time (SAM) avg: 0.39, (0.39, 0.39)\n",
      "\tReading IO time (Reference Genome) avg: 0.01, (0.01, 0.01)\n",
      "\tIndex read time avg: 0.02, (0.02, 0.02)\n",
      "\n",
      "\tOverall time (sec) (Excluding Index reading time):\n",
      "\tPROCESS() (Total compute time + (read + SAM) IO time) : 3.57\n",
      "\tMEM_PROCESS_SEQ() (Total compute time (Kernel + SAM)), avg: 1.45, (1.45, 1.45)\n",
      "\n",
      "\t SAM Processing time (sec):\n",
      "\t--WORKER_SAM avg: 0.19, (0.19, 0.19)\n",
      "\n",
      "\tKernels' compute time (sec):\n",
      "\tTotal kernel (smem+sal+bsw) time avg: 1.26, (1.26, 1.26)\n",
      "\t\tSMEM compute avg: 0.40, (0.44, 0.37)\n",
      "\t\tSAL compute avg: 0.08, (0.10, 0.07)\n",
      "\t\t\t\tMEM_SA avg: 0.07, (0.08, 0.06)\n",
      "\n",
      "\t\tBSW time, avg: 0.68, (0.69, 0.66)\n",
      "\n",
      "Important parameter settings: \n",
      "\tBATCH_SIZE: 512\n",
      "\tMAX_SEQ_LEN_REF: 256\n",
      "\tMAX_SEQ_LEN_QER: 128\n",
      "\tMAX_SEQ_LEN8: 128\n",
      "\tSEEDS_PER_READ: 500\n",
      "\tSIMD_WIDTH8 X: 32\n",
      "\tSIMD_WIDTH16 X: 16\n",
      "\tAVG_SEEDS_PER_READ: 64\n",
      "Looking to launch executable \"/usr/local/bin/bwa-mem2.avx2\", simd = .avx2\n",
      "Launching executable \"/usr/local/bin/bwa-mem2.avx2\"\n",
      "-----------------------------\n",
      "Executing in AVX2 mode!!\n",
      "-----------------------------\n",
      "* SA compression enabled with xfactor: 8\n",
      "* Ref file: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Entering FMI_search\n",
      "* Index file found. Loading index from /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta.bwt.2bit.64\n",
      "* Reference seq len for bi-index = 9950285\n",
      "* sentinel-index: 7387327\n",
      "* Count:\n",
      "0,\t1\n",
      "1,\t2386455\n",
      "2,\t4975143\n",
      "3,\t7563831\n",
      "4,\t9950285\n",
      "\n",
      "* Reading other elements of the index from files /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Index prefix: /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "* Read 0 ALT contigs\n",
      "* Done reading Index!!\n",
      "* Reading reference genome..\n",
      "* Binary seq file = /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta.0123\n",
      "* Reference genome size: 9950284 bp\n",
      "* Done reading reference genome !!\n",
      "\n",
      "------------------------------------------\n",
      "1. Memory pre-allocation for Chaining: 4458.8284 MB\n",
      "2. Memory pre-allocation for BSW: 7667.7448 MB\n",
      "3. Memory pre-allocation for BWT: 2474.0536 MB\n",
      "------------------------------------------\n",
      "* Threads used (compute): 32\n",
      "* No. of pipeline threads: 2\n",
      "\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 88574092, nseq: 325457\n",
      "\t[0000][ M::kt_pipeline] read 325457 sequences (88574092 bp)...\n",
      "[0000] Calling mem_process_seqs.., task: 0\n",
      "[0000] 1. Calling kt_for - worker_bwt\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 0, nseq: 0\n",
      "[0000] 2. Calling kt_for - worker_aln\n",
      "[0000] 3. Calling kt_for - worker_sam\n",
      "\t[0000][ M::mem_process_seqs] Processed 325457 reads in 72.309 CPU sec, 2.378 real sec\n",
      "[0000] read_chunk: 320000000, work_chunk_size: 0, nseq: 0\n",
      "[0000] Computation ends..\n",
      "No. of OMP threads: 32\n",
      "Processor is running @1995.515520 MHz\n",
      "Runtime profile:\n",
      "\n",
      "\tTime taken for main_mem function: 4.90 sec\n",
      "\n",
      "\tIO times (sec) :\n",
      "\tReading IO time (reads) avg: 2.00, (2.00, 2.00)\n",
      "\tWriting IO time (SAM) avg: 0.37, (0.37, 0.37)\n",
      "\tReading IO time (Reference Genome) avg: 0.01, (0.01, 0.01)\n",
      "\tIndex read time avg: 0.02, (0.02, 0.02)\n",
      "\n",
      "\tOverall time (sec) (Excluding Index reading time):\n",
      "\tPROCESS() (Total compute time + (read + SAM) IO time) : 4.86\n",
      "\tMEM_PROCESS_SEQ() (Total compute time (Kernel + SAM)), avg: 2.38, (2.38, 2.38)\n",
      "\n",
      "\t SAM Processing time (sec):\n",
      "\t--WORKER_SAM avg: 0.60, (0.60, 0.60)\n",
      "\n",
      "\tKernels' compute time (sec):\n",
      "\tTotal kernel (smem+sal+bsw) time avg: 1.78, (1.78, 1.78)\n",
      "\t\tSMEM compute avg: 0.50, (0.56, 0.47)\n",
      "\t\tSAL compute avg: 0.07, (0.07, 0.06)\n",
      "\t\t\t\tMEM_SA avg: 0.05, (0.06, 0.04)\n",
      "\n",
      "\t\tBSW time, avg: 1.10, (1.13, 1.07)\n",
      "\n",
      "Important parameter settings: \n",
      "\tBATCH_SIZE: 512\n",
      "\tMAX_SEQ_LEN_REF: 256\n",
      "\tMAX_SEQ_LEN_QER: 128\n",
      "\tMAX_SEQ_LEN8: 128\n",
      "\tSEEDS_PER_READ: 500\n",
      "\tSIMD_WIDTH8 X: 32\n",
      "\tSIMD_WIDTH16 X: 16\n",
      "\tAVG_SEEDS_PER_READ: 64\n"
     ]
    }
   ],
   "source": [
    "polypolish_dirname = outpath/'polypolish'\n",
    "polypolish_dirname.mkdir(exist_ok=True)\n",
    "\n",
    "!bwa-mem2 index {outpath/'final_consensus.fasta'}\n",
    "\n",
    "!bwa-mem2 mem -t {threads} -a {outpath/'final_consensus.fasta'} {SR1} > {polypolish_dirname/'alignments_1.sam'}\n",
    "!bwa-mem2 mem -t {threads} -a {outpath/'final_consensus.fasta'} {SR2} > {polypolish_dirname/'alignments_2.sam'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e20bc0dd-2d56-4411-ace2-2ec7b6e22393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mPolypolish insert size alignment filter\u001b[0m \u001b[2m(2022-01-27 15:26:00)\u001b[0m\n",
      "\u001b[2m    This script is a pre-processing filter that can be run on SAM alignments\u001b[0m\n",
      "\u001b[2mbefore they are used in Polypolish. It looks at each read pair and flags\u001b[0m\n",
      "\u001b[2malignments that do not seem to be part of a concordant pair. This can improve\u001b[0m\n",
      "\u001b[2mthe accuracy of Polypolish, especially near the edges of repeats.\u001b[0m\n",
      "\n",
      "Input alignments:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_1.sam\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_2.sam\n",
      "\n",
      "Output alignments:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_1.sam\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_2.sam\n",
      "\n",
      "Settings:\n",
      "  --orientation auto\n",
      "  --low 0.1\n",
      "  --high 99.9\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mLoading alignments\u001b[0m \u001b[2m(2022-01-27 15:26:00)\u001b[0m\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_1.sam: 355,092 alignments from 325,359 reads\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_2.sam: 355,545 alignments from 325,412 reads\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinding insert size thresholds\u001b[0m \u001b[2m(2022-01-27 15:26:07)\u001b[0m\n",
      "\u001b[2m    Read pairs with exactly one alignment per read are used to determine the\u001b[0m\n",
      "\u001b[2morientation and insert size thresholds for the read set.\u001b[0m\n",
      "\n",
      "fr: 312699 pairs\n",
      "rf: 2247 pairs\n",
      "ff: 2 pairs\n",
      "rr: 2 pairs\n",
      "\n",
      "Automatically determined correct orientation: fr\n",
      "\n",
      "Low threshold:  43 (0.1st percentile)\n",
      "High threshold: 841 (99.9th percentile)\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFiltering SAM files\u001b[0m \u001b[2m(2022-01-27 15:26:09)\u001b[0m\n",
      "\u001b[2m    Read alignments that are part of a good pair (correct orientation and\u001b[0m\n",
      "\u001b[2minsert size) pass the filter and are written unaltered to the output file. Read\u001b[0m\n",
      "\u001b[2malignments which are not part of good pair are written to the output file with\u001b[0m\n",
      "\u001b[2ma \"ZP:Z:fail\" tag so Polypolish will not use them.\u001b[0m\n",
      "\n",
      "Filtering /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_1.sam:\n",
      "  351,259 pass\n",
      "  3,931 fail\n",
      "\n",
      "Filtering /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/alignments_2.sam:\n",
      "  351,316 pass\n",
      "  4,274 fail\n",
      "\n",
      "\n",
      "\u001b[93m\u001b[1m\u001b[4mFinished!\u001b[0m \u001b[2m(2022-01-27 15:26:17)\u001b[0m\n",
      "Alignments before filtering: 710,637\n",
      "Alignments after filtering:  702,575\n",
      "\n",
      "Time to run: 0:00:16.877152\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!polypolish_insert_filter.py \\\n",
    "--in1 {polypolish_dirname/'alignments_1.sam'} --in2 {polypolish_dirname/'alignments_2.sam'} \\\n",
    "--out1 {polypolish_dirname/'filtered_1.sam'} --out2 {polypolish_dirname/'filtered_2.sam'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4e02d76d-7233-401b-acfd-b04bbdb2c3ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[1;4;93mStarting Polypolish\u001b[0m \u001b[2m(2022-01-27 15:26:18)\u001b[0m\n",
      "\u001b[2m    Polypolish is a tool for polishing genome assemblies with short reads.\n",
      "Unlike other tools in this category, Polypolish uses SAM files where each read\n",
      "has been aligned to all possible locations (not just a single best location).\n",
      "This allows it to repair errors in repeat regions that other alignment-based\n",
      "polishers cannot fix.\u001b[0m\n",
      "\n",
      "Polypolish version: 0.5.0\n",
      "\n",
      "Input assembly:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/final_consensus.fasta\n",
      "\n",
      "Input short-read alignments:\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_1.sam\n",
      "  /media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_2.sam\n",
      "\n",
      "Settings:\n",
      "  --fraction_invalid 0.2\n",
      "  --fraction_valid 0.5\n",
      "  --max_errors 10\n",
      "  --min_depth 5\n",
      "  not logging debugging information\n",
      "\n",
      "\n",
      "\u001b[1;4;93mLoading assembly\u001b[0m \u001b[2m(2022-01-27 15:26:18)\u001b[0m\n",
      "cluster_001_consensus (4,975,142 bp)\n",
      "\n",
      "\n",
      "\u001b[1;4;93mLoading alignments\u001b[0m \u001b[2m(2022-01-27 15:26:18)\u001b[0m\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_1.sam: 355,092 alignments from 325,359 reads\n",
      "/media/Central_Lab_Storage/MinION_2022/20220126_Sal/tryclcyer/barcode03/polypolish/filtered_2.sam: 355,545 alignments from 325,412 reads\n",
      "\n",
      "Filtering for high-quality end-to-end alignments:\n",
      "  594,721 alignments kept\n",
      "  115,916 alignments discarded\n",
      "\n",
      "\n",
      "\u001b[1;4;93mPolishing assembly sequences\u001b[0m \u001b[2m(2022-01-27 15:26:25)\u001b[0m\n",
      "\u001b[2m    For each position in the assembly, Polypolish determines the read depth\n",
      "at that position and collects all aligned bases. It then polishes the assembly\n",
      "by looking for positions where the pileup unambiguously supports a different\n",
      "sequence than the assembly.\u001b[0m\n",
      "\n",
      "Polishing cluster_001_consensus (4,975,142 bp):\n",
      "  mean read depth: 29.8x\n",
      "  35 bp have a depth of zero (99.9993% coverage)\n",
      "  144 positions changed (0.0029% of total positions)\n",
      "  estimated pre-polishing sequence accuracy: 99.9971%\n",
      "\n",
      "\n",
      "\u001b[1;4;93mFinished!\u001b[0m \u001b[2m(2022-01-27 15:26:27)\u001b[0m\n",
      "Polished sequence (to stdout):\n",
      "  cluster_001_consensus_polypolish (4,975,234 bp)\n",
      "\n",
      "Time to run: 0:00:08.961873\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!polypolish \\\n",
    "{outpath/'final_consensus.fasta'} \\\n",
    "{polypolish_dirname/'filtered_1.sam'} {polypolish_dirname/'filtered_2.sam'} > {outpath/'polished.fasta'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5fc92e3-54c1-47a0-8b66-d34ed2868e65",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
